haskell/experimental/private/module.bzl

```diff
@@ -361,29 +418,36 @@ def build_haskell_modules(ctx, hs, cc, posix, package_name, with_shared, hidir,
             dependencies
     """
 
+    per_module_transitive_interfaces = _merge_narrowed_deps_dicts(ctx.label, ctx.attr.narrowed_deps).per_module_transitive_interfaces
+
     dep_info = gather_dep_info(ctx.attr.name, ctx.attr.deps)
-    transitive_module_deps = _reorder_module_deps_to_postorder(ctx.label, ctx.attr.modules)
+    narrowed_deps_info = gather_dep_info(ctx.attr.name, ctx.attr.narrowed_deps)
+    transitive_module_deps = _reorder_module_deps_to_postorder(ctx.label, ctx.attr.modules, per_module_transitive_interfaces)
     module_outputs = {dep.label: _declare_module_outputs(hs, with_shared, hidir, odir, dep) for dep in transitive_module_deps}
+    package_ids = hs.package_ids + all_dependencies_package_ids(ctx.attr.narrowed_deps)
 
     module_interfaces = {}
     module_objects = {}
     for dep in transitive_module_deps:
         his, os = _collect_module_outputs_of_direct_deps(with_shared, module_outputs, dep)
         interface_inputs = _collect_module_inputs(module_interfaces, his, dep)
         object_inputs = _collect_module_inputs(module_objects, os, dep)
+        narrowed_interfaces = _collect_narrowed_deps_module_files(per_module_transitive_interfaces, dep)
 
         _build_haskell_module(
             ctx,
             hs,
             cc,
             posix,
             dep_info,
+            narrowed_deps_info,
```

Comment:
Are they used separately, or could they be passed in merged form same as `interface_inputs`?

Reply:
They are used separately. All of the interface files in `dep_info` are passed as inputs, while only a few from `narrowed_deps_info` are passed.

Reply:
Makes sense, thanks for clarifying.